### DISCUSSION TITLE
Ways to do Typing

### DISCUSSION THREAD

**Discussion by Moritz-Schmidt** [wmalgadey/PyTado#132] [Ideas]
There are multiple ways to type this library. 

#### 1. Mashumaro (or other serialization/deserialization library)
@erwindouna started implementing [mashumaro](https://github.com/Fatal1ty/mashumaro) models in his rewrite. These could be used to add typing to the HTTP API responses and the return types of the PyTado functions. 
This has the drawback that it would probably break backwards compatibility since the functions will return a dataclass instead of a dict. This may be avoided by either implementing functions like `__getitem__` so the dataclass also acts like a dict (however I'm not sure if this will work properly) or by using mashumaro just for validation and still returning a dict.

#### 2. TypedDict 
The Models mentioned above could also be converted to TypedDict's which would be great for backwards compatibility but not as nice for readability and validation.

Personally I'd favour Mashumaro since the Models already mostly exist. I think for backwards compatibility returning dicts in the Tado Class in interface.py would be good. The new methods should return mashumaru objects.

What are other opinions on this? I can start working on typing once we decided how it should be done.

**Comment by wmalgadey** [comment]
What about typing inside out? So we use strict typing everywhere we can inside the library but do not rely on heavy typing at the outside interfaces, like data classes? Much like an API, where you have to separate the typing into a definition file (swagger/openapi)

**Reply by Moritz-Schmidt** [reply]
We could use mashumaru internally to deserialize and validate the HTTP Responses and return (typed) dicts? However, IMO it'd be easier and better to return mashumaru objects (which are just dataclasses with added functionality to parse JSON into a dataclass and vice versa). 

As an example for what I have in mind:

**models.py:**
```py
@dataclass
class GetMe(DataClassORJSONMixin):
    """GetMe model represents the user's profile information."""

    name: str
    email: str
    id: str
    username: str
    locale: str
    homes: list[Home]
    
    class Config(BaseConfig):
        serialize_by_alias = True
```

`serialize_by_alias = True` ensures that 
```py
self._http.request(request) == GetMe.from_dict(self._http.request(request)).to_dict()
``` 
for attributes which are camelCase named in the HTTP response but snake_case in the dataclass.


**base_tado.py:**
```py
    def get_me(self) -> GetMe:
        """Gets home information."""

        request = TadoRequest()
        request.action = Action.GET
        request.domain = Domain.ME

        return GetMe.from_dict(self._http.request(request))
 ```
 
 and for backwards compatibility in interface.py:
 ```py
    @deprecated("get_me")
    def getMe(self) -> dict[str, Any]:
        """Gets home information. (deprecated)"""
        return self.get_me().to_dict()
  ```

**Reply by erwindouna** [reply]
Nice idea, @Moritz-Schmidt! There are quite some other parts which have camel case. If we go fully for the PEP8 standard, we might as well use the activity to update.
I like your approach to keep the camel case as a backward compatibility. This saves loads of refactoring of already running instances (ie Home Assistant).

**Reply by wmalgadey** [reply]
Could we use the data from #118?

**Reply by Moritz-Schmidt** [reply]
We could as there are tools for generating dataclasses from openAPI specs but I don't really see the point since the dataclasses already mostly exist in https://github.com/erwindouna/python-tado and for Tado X I've already started writing them.

Or did you mean something else?

**Reply by wmalgadey** [reply]
I like generating my popo classes 😄

**Reply by Moritz-Schmidt** [reply]
I don't see any reason to use plain old python objects (I think that's what "popo" means?). Python standard library already has dataclasses, so this style of defining data-only objects is quite well known and much nicer to use. Since we also have a lot of the work already done, I would much prefer using Mashumaro (or Pydantic, which would be quite a simple change for the existing mashumaro code, and is the more popular project).

I'm against spending a lot of effort on a more manual solution where we don't use any library for deserialisation and validation of JSON responses.

**Reply by erwindouna** [reply]
That's probably what popo means, whereas Wolfgang is most likely just joking a bit. :)
Pydantic would be the better choice now. It has a better ecosystem behind it and is more popular nowadays.

**Reply by wmalgadey** [reply]
Yes just a bit joking, I'm German and POPO means butt for little kids 😄, but I still use "plain old ... objects" a lot, because I like the idea of lightweight objects, which can be de-/serialized easily. But I think this is not necessary for python environments actually.

**Reply by Moritz-Schmidt** [reply]
alright, then I'll start working on doing it with pydantic :)

**Comment by wmalgadey** [comment]
@Moritz-Schmidt can we help you somehow?

**Comment by wmalgadey** [comment]
we merged the changes from @Moritz-Schmidt to main. I will close this discussion.
